Improve :has() pseudo-class documentation (#20889)
mdn/content: files/en-us/web/css/_colon_has/index.md

{{CSSRef}}

The functional **`:has()`** CSS [pseudo-class](/en-US/docs/Web/CSS/Pseudo-classes) represents an element if any of the [relative selectors](/en-US/docs/Web/CSS/CSS_Selectors#relative_selector) that are passed as an argument match at least one element when anchored against this element. This pseudo-class presents a way of selecting a parent element or a previous sibling element with respect to a reference element by taking a [forgiving relative selector list](/en-US/docs/Web/CSS/Selector_list#forgiving_relative_selector_list) as an argument.

```css
/* Selects an h1 heading with a
paragraph element that immediately follows
the h1 and applies the style to h1 */
h1:has(+ p) { margin-bottom: 0; }

> ```

Reviewer: Is this intended to be replaced with an interactive example later? In general I think static examples should live in the "Examples" section.

Author: Hi @Josh-Cena, here I have followed the [template for selector pages](https://developer.mozilla.org/en-US/docs/MDN/Writing_guidelines/Page_structures/Page_types/CSS_selector_page_template), where it mentions using "/* Insert code block showing common use cases */" right before the "Syntax" section.

For example, see these pages [`:first`](https://developer.mozilla.org/en-US/docs/Web/CSS/:first) and [`:dir()`](https://developer.mozilla.org/en-US/docs/Web/CSS/:dir). 

But then there's also this format in [`:hover`](https://developer.mozilla.org/en-US/docs/Web/CSS/:hover). Maybe when we add an interactive example to `:has()`, this static example can be moved down to "Examples".

## Syntax

```
:has( <forgiving-relative-selector-list> )
```

> > **Note:** The `:has()` pseudo-class cannot be nested within another `:has()`. Also, pseudo-elements are not valid selectors within `:has()`. This is because many pseudo-elements exist conditionally based on the styling of their ancestors and allowing these to be queried by `:has()` can introduce cyclic querying.

Reviewer 2: ```suggestion
The `:has()` pseudo-class cannot be nested within another `:has()`. Also, pseudo-elements are not valid selectors within `:has()`. This is because many pseudo-elements exist conditionally based on the styling of their ancestors and allowing these to be queried by `:has()` can introduce cyclic querying.

The relative selector list is [forgiving](https://drafts.csswg.org/selectors-4/#typedef-forgiving-selector-list). Normally in CSS when a selector in a selector list is invalid then the whole list is deemed invalid. When a selector in a `:has()` selector list fails to parse, the incorrect or unsupported selector will be ignored and the others used. Note that if the `:has()` itself is not supported in a browser, the entire selector block will fail (unless the :has() itself is in a forgiving selector list such as is found in [`:is()`](/en-US/docs/Web/CSS/:is) and [`:where()`](/en-US/docs/Web/CSS/:where).

While `:has()` cannot be nested in a `:has()`, as the selector list is forgiving, it will just be ignored. The selector will not fail. The same holds true for pseudo-elements.
```

Author: - Any reason for not having that text as "Note:"?
- Actually, the first paragraph on the page links to the definition of the 'forgiving selector list':
"This pseudo-class presents a way of selecting a parent element or a previous sibling element with respect to a reference element by taking a [forgiving relative selector list](https://pr20889.content.dev.mdn.mozit.cloud/en-US/docs/Web/CSS/Selector_list#forgiving_relative_selector_list) as an argument."

In this PR, I also submitted the updated 'Selector list' page to update those definitions. Do we need to explain it here again?

Reviewer 2: The one note became a 3 paragraph explanation, so didn't want to notify it. As the list is forgiving including `:has()` or `::pseudoelement` will not break anything, with the why in the next two paragraphs, I wanted to keep all three on the same level. 

While it does link to the forgiving selector list and to the relative selector, many people don't know what these terms mean. We don't want 100% of people to need to leave the page to understand its contents. Many of the terms can be linked to; but I strongly feel these need to be included in this document.

Author: I feel if the same concept is explained on different pages, it will lead to maintenance issues. The explanation of the behavior and examples can live on one page and others can link to it. The fact that folks don't know about these terms is relevant only now and but hopefully, more people will understand these terms in the future, in which case it will make sense for the concepts to live on a single page.

Currently, [`:where()`](https://developer.mozilla.org/en-US/docs/Web/CSS/:where#forgiving_selector_parsing) and [`:is()`](https://developer.mozilla.org/en-US/docs/Web/CSS/:is#forgiving_selector_parsing) each have a section on "Forgiving selector parsing", perhaps because it was a new concept at the time. But that has now resulted in a repition of the same content, which leads to maintenance issues - if it is in one place, only one place would need to be updated in the future if there's any change in behavior or examples need to be updated. I wanted to update `:is()` and `:where()` via a separate PR.

## Examples

### With the sibling combinator

> The `:has()` rule in the following example adjusts the spacing after `H1` headings if they are immediately followed by an `H2` heading.

Reviewer 2: ```suggestion
The `:has()` style declaration in the following example adjusts the spacing after `H1` headings if they are immediately followed by an `H2` heading.
```

#### HTML

```html
<section>
  <article>
    <h1>Morning Times</h1>
    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.</p>
  </article>
  <article>

>     <h1>Morning Times</h1>

Reviewer 2: maybe call this afternoon times so you can reference and say the afternoon times heading is matched.

    <h2>Delivering you news every morning</h2>
    <p>Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.</p>
  </article>
</section>
```

#### CSS

```css hidden
section {
    display: flex;
    align-items: start;
    justify-content: space-around;
}

article {
  display: inline-block;
  width: 40%;
}

h1, h2 {
  font-size: 1.2em;
}

h2 {
  font-size: 1.0em;
  color: rgb(150, 149, 149);
}
```

```css
h1, h2 {
  margin: 0 0 1.0rem 0;
}

h1:has(+ h2) {
  margin: 0 0 0.25rem 0;
}
```

#### Result

{{EmbedLiveSample('With_the_sibling_combinator', 600, 150)}}

> This example shows two similar texts side-by-side for comparison – the left one with an `H1` heading followed by a paragraph and the right one with an `H1` heading followed by an `H2` heading and then a paragraph. In the example on the right, `:has()` helps to select the `H1` element that is immediately followed by (indicated by [`+`](/en-US/docs/Web/CSS/Adjacent_sibling_combinator)) an `H2` element and the CSS rule reduces the spacing after such an `H1` element. Without the `:has()` pseudo-class, you cannot select the preceding sibling or parent element.

Reviewer 2: you can actually select a preceding sibling if it is of the same type. I showed this with the fieldset example explanation

Author: How would you recommend to fix this? Use `<p>` instead of `<h2>`?

Reviewer 2: fix in comment below.

### With the :not() pseudo-class

The form in this example contains three `<fieldset>` groups. The first two are indicated as mandatory and are styled pink when no option is selected. In addition, when no option is selected in the first two `<fieldset>` groups, an asterisk displays after the question. The `:has()` pseudo-class is used here to style the form elements based on the state of the child or sibling elements.

#### HTML

```html
<form>
  <p>Fill out our questionnaire:</p>
  <fieldset class="required">
    <legend>Which activity are you interested in?</legend>
    <input type="radio" name="activity" id="activity1">
    <label for="activity1">Running</label> 
    <input type="radio" name="activity" id="activity2">
    <label for="activity2">Rock climbing</label>
    <input type="radio" name="activity" id="activity3">
    <label for="activity3">Swimming</label>
  </fieldset>

>   <fieldset class="required">

Reviewer 2: if this is required, it has to be required: the screen reader user has no way of knowing it's required. And, the browser doesn't know either.

    <legend>What days work for you?</legend>
    <input type="checkbox" id="day1">
    <label for="day1">Monday</label> 
    <input type="checkbox" id="day2">
    <label for="day2">Wednesday</label>
    <input type="checkbox" id="day3">
    <label for="day3">Friday</label>
  </fieldset>

  <fieldset>
    <legend>Join our newsletter</legend>

>     <input type="radio" name="newsletter" id="newsweekly">

Author: I am removing this form example for now. It seems we need a better example that correctly reflects the importance of `:has()` while also using other coding best practices. We can work on an example offline and then add it to the docs.

    <label for="newsweekly">Weekly</label> 
    <input type="radio" name="newsletter" id="newsmonthly">
    <label for="newsmonthly">Monthly</label> 
  </fieldset>
</form>
```

#### CSS

```css hidden
form { 
    border: 4px solid white;
    width: max-content;
    font-family: sans-serif;
    margin-left: 2em;
  }

legend {
    background: #7B0D1E;
    color: #fff;
    padding: 3px 6px;
    background: white;
    color: black;
    font-weight: bold;
}
```

```css

fieldset:has(+ fieldset) {
  margin-bottom: 1em;
}

fieldset.required:not(:has(input:checked)) {
  background: pink;
}

fieldset.required:not(:has(input:checked)) legend::after {

>   content: " ✳︎";

Reviewer 2: ```suggestion
  content: " (required)";
```

Reviewer 2: i am not sure if adding content is ok with CSS, or if we should also had an aria attribute

Reviewer 2: we could teach :has() and a11y at the same time.

```html
<legend>Label here <span aria-label="required">*</span></legend>
```

```css 
fieldset:has( > legend > [aria-label="required"])
```

That shows ho to correctly require a checkbox group and how to use the child selector in a has.

>   content: " ✳︎";

Reviewer 2: This is not best practices. Instead use:

```html
<fieldset>
   <legend>What days work for you? <span aria-label="required">*</span></legend>
```

and then in CSS you can teach:

```css
fieldset:has( > legend > [aria-label="required"])
/* or */
legend:has( > [aria-label="required"])
```

also, since fieldsets can be nested, this example, as is, may or may not be robust. Plus you don't want to toggle the generated content based on the state of an input. That would be weird.

  color: red;
}
```

#### Result

{{EmbedLiveSample('With_the_:not()_pseudo-class', 600, 300)}}

- The first `:has()` pseudo-class selects any `<fieldset>` element adjacent to another `<fieldset>` element and applies a bottom spacing to the preceding `<fieldset>` element. This selects the first and second `<fieldset>` elements, and the spacing is applied after them.

> - The second `:has()` pseudo-class in combination with the [`:not()`](/en-US/docs/Web/CSS/:not) pseudo-class selects any `<fieldset>` element with the `required` class containing an input radio or checkbox that hasn't been toggled to an [on state](/en-US/docs/Web/CSS/:checked). This selects the first and second `<fieldset>` elements, and applies the pink background to the `<fieldset>` elements.

Reviewer 2: let's add a 1px red outline as well, so color is not the only thing changed. Outline, not border, so the box model isn't altered.

- The third `:has()` pseudo-class selects the same elements as the second `:has()` pseudo-class, but in this case the `<legend>` element is styled to add the asterisk after the question.

- After you select an option in the first two groups, the `:has()` selection no longer applies and the initial formatting and asterisk are removed.

Be careful about the order of `:has()` and [`:not()`](/en-US/docs/Web/CSS/:not) pseudo-classes because swapping the order does not select the same elements. For example, in the following rule, a [selector list](/en-US/docs/Web/CSS/Selector_list) is passed as an argument to `:has()`. The `:has()` and `:not()` pseudo-classes here will select only those `<section>` elements that don't contain any heading elements.

```css
section:not(:has(h1, h2, h3, h4, h5, h6)) {...}
```

However, if you swap the order of the two pseudo-classes as shown below:

```css
section:has(:not(h1, h2, h3, h4, h5, h6)) {...}
```

> The `:not()` pseudo-class will select any element that's not a heading element, such as `<p>`, `<a>`, or `<button>` and pass that as an argument to `:has()`, which in turn, will select any `<section>` element that contains those other non-heading elements. This might eventually end up selecting even those `<section>` elements that do have a heading element along with other non-heading elements.

Reviewer 2: ```suggestion
The `:not()` pseudo-class will select any element that's not a heading element, such as `<p>`, `<a>`, or `<button>` and pass that as an argument to `:has()`, which in turn, will select any `<section>` element that contains those other non-heading elements. This matches every section that contains a single element that is not a heading, meaning it will match a section that contains both paragraphs and headings. For this reason, it is recommended to always use explicit, specific combinators in all components `:has()` within the relational selector list parameter.
```

## See also

- [`:is()`](/en-US/docs/Web/CSS/:is), [`:not()`](/en-US/docs/Web/CSS/:not)

> - [CSS selectors](/en-US/docs/Web/CSS/CSS_Selectors)

Author: I used the learn link. Thanks!

## See also

- [`:is()`](/en-US/docs/Web/CSS/:is), [`:not()`](/en-US/docs/Web/CSS/:not)
- [CSS selectors](/en-US/docs/Web/CSS/CSS_Selectors)
- [Selector list](/en-US/docs/Web/CSS/Selector_list)
- [Locating DOM elements using selectors](/en-US/docs/Web/API/Document_object_model/Locating_DOM_elements_using_selectors)

> - For more inspiring ideas, see [Using :has() as a CSS parent selector and much more](https://webkit.org/blog/13096/css-has-pseudo-class/#using-has-with-the-child-combinator).

Reviewer 2: ```suggestion
```
